### PR TITLE
[WIP] React on rails from dummy

### PR DESCRIPTION
**Pull request by kpuputti**

### CODE REVIEW

File: config/username_blacklist.yml [hunk 1]
```diff
@@ -80,6 +80,7 @@
 - ssladministrator
 - sslwebmaster
 - statistics
+- styleguide
```
Comment (Gnito) [inline config/username_blacklist.yml]: We should check this again from db before merging

File: app/views/layouts/styleguide.erb [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+<head>
+  <title>Styleguide with Hot Module Replacement HMR</title>
+
+  <%= env_stylesheet_link_tag(static: 'application',
+                              hot: 'application_hot',
+                              media: 'all') %>
+
+  <!-- This is to load the hot assets. -->
+  <%= env_javascript_include_tag(hot: ['http://localhost:3500/vendor-bundle.js',
+                                       'http://localhost:3500/app-bundle.js']) %>
```
Comment (Gnito) [inline app/views/layouts/styleguide.erb]: Get Rails domain from somewhere

File: app/assets/stylesheets/main.css.scss [hunk 3]
```diff
@@ -5,6 +5,7 @@
 @import "compass/css3/border-radius";
 @import "compass/css3/box-shadow";
 @import "mixins/all";
+// @import 'app-bundle';
```
Comment (Gnito) [inline app/assets/stylesheets/main.css.scss]: uncomment

File: app/controllers/styleguide/pages_controller.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  rescue_from ReactOnRails::PrerenderError do |err|
+    Rails.logger.error(err.message)
+    Rails.logger.error(err.backtrace.join("\n"))
+    redirect_to styleguide_path,
+                flash: { error: "Error prerendering in react_on_rails. See server logs." }
+  end
+
+  private
+
+  def initialize_shared_store
```
Comment (Gnito) [inline app/controllers/styleguide/pages_controller.rb]: Remove?

File: app/views/layouts/styleguide.erb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  <%= env_javascript_include_tag(static: 'application',
+                                 hot: 'application_hot') %>
+
+  <%= csrf_meta_tags %>
+</head>
+<body>
+<% flash.each do |key, value| %>
```
Comment (Gnito) [inline app/views/layouts/styleguide.erb]: What does this do?

File: app/views/layouts/styleguide.erb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  <%= content_tag :div, value, class: "flash #{key}" %>
+<% end %>
+
+<%= yield %>
+
+<!-- This is a placeholder for ReactOnRails to know where to render the store props for
+     client side hydration -->
+<%= redux_store_hydration_data %>
```
Comment (Gnito) [inline app/views/layouts/styleguide.erb]: Remove?

File: client/.eslintrc [hunk 7]
```diff
@@ -0,0 +1,48 @@
+---
```
Comment (Gnito) [inline client/.eslintrc]: @kpuputti check this out?

File: client/.jscsrc [hunk 8]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (Gnito) [inline client/.jscsrc]: Remove?

File: client/README.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
+Please see parent directory README.md.
```
Comment (Gnito) [inline client/README.md]: Rewrite

File: client/app/components/RailsContext.jsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import React, { PropTypes } from 'react';
+import _ from 'lodash';
+
+// This helps to figure out and debug context passed from Rails side
+function renderContextRows(railsContext) {
```
Comment (Gnito) [inline client/app/components/RailsContext.jsx]: Remove visual presentation of railsContext?

File: client/app/startup/ExampleServerReduxApp.jsx [hunk 11]
```diff
@@ -0,0 +1,37 @@
+// This is loaded by execJs and Rails to generate the HTML used for server rendering.
```
Comment (Gnito) [inline client/app/startup/ExampleServerReduxApp.jsx]: Remove one of these app registrations (client / server)

File: client/package.json [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    "postcss-loader": "0.8.2",
+    "react": "0.14.8",
+    "react-dom": "0.14.8",
+    "react-on-rails": "5.2.0",
+    "react-redux": "4.4.2",
+    "react-router": "2.0.1",
+    "redux": "3.4.0",
+    "redux-thunk": "2.0.1",
+    "resolve-url-loader": "1.4.3",
+    "sass-loader": "3.2.0",
```
Comment (Gnito) [inline client/package.json]: Remove?

File: client/package.json [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    "body-parser": "1.15.0",
+    "babel-eslint": "6.0.2",
+    "eslint": "2.7.0",
+    "eslint-config-airbnb": "6.2.0",
+    "eslint-plugin-react": "4.3.0",
+    "react-transform-hmr": "1.0.4",
+    "webpack-dev-server": "1.14.1"
+  },
+  "browser": {
+    "fs": false
```
Comment (Gnito) [inline client/package.json]: WTF?

File: client/webpack.client.base.config.js [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+    // This will contain the app entry points
+    app: [
+      './app/startup/clientRegistration',
+    ],
+  },
+  resolve: {
+    extensions: ['', '.js', '.jsx'],
+    alias: {
+      libs: path.join(process.cwd(), 'app', 'libs'),
```
Comment (Gnito) [inline client/webpack.client.base.config.js]: Maybe remove

File: client/webpack.server.rails.build.config.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    './app/startup/serverRegistration',
+  ],
+  output: {
+    filename: 'server-bundle.js',
+    path: '../app/assets/webpack',
+  },
+  resolve: {
+    extensions: ['', '.js', '.jsx'],
+    alias: {
+      libs: path.join(process.cwd(), 'app', 'libs'),
```
Comment (Gnito) [inline client/webpack.server.rails.build.config.js]: Maybe remove

File: config/initializers/assets.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+# Add folder with webpack generated assets to assets.paths
+Rails.application.config.assets.paths << Rails.root.join("app", "assets", "webpack")
+
+# Precompile additional assets.
+# application.js, application.css, and all non-JS/CSS in app/assets folder are already added.
+Rails.application.config.assets.precompile << "server-bundle.js"
+
+Rails.application.config.assets.precompile +=
+  if ENV["REACT_ON_RAILS_ENV"] == "HOT"
+    ["application_hot.js", "application_hot.css"]
```
Comment (Gnito) [inline config/initializers/assets.rb]: remove hot.js?

File: config/initializers/react_on_rails.rb [hunk 17]
```diff
@@ -0,0 +1,59 @@
+module RenderingExtension
+  # Return a Hash that contains custom values from the view context that will get passed to
+  # all calls to react_component and redux_store for rendering
+  def self.custom_context(view_context)
+    {
+      somethingUseful: view_context.session[:something_useful]
```
Comment (Gnito) [inline config/initializers/react_on_rails.rb]: Comment away (but keep)

**Comment by Gnito** [comment]
Cucumber tests fail because there's no vendor-bundle!

File: app/views/layouts/styleguide.erb (hunk 2 above)
Comment (Gnito) [inline app/views/layouts/styleguide.erb]: I'll start with this one - vesa

File: app/controllers/styleguide/pages_controller.rb (hunk 4 above)
Comment (Gnito) [inline app/controllers/styleguide/pages_controller.rb]: I'll check this one

File: app/views/layouts/styleguide.erb (hunk 6 above)
Comment (Gnito) [inline app/views/layouts/styleguide.erb]: I'll check this one

File: app/assets/stylesheets/main.css.scss (hunk 3 above)
Comment (Gnito) [inline app/assets/stylesheets/main.css.scss]: I'll take this one

File: client/.jscsrc (hunk 8 above)
Comment (Gnito) [inline client/.jscsrc]: I'll take this one

File: client/app/components/RailsContext.jsx (hunk 10 above)
Comment (Gnito) [inline client/app/components/RailsContext.jsx]: I'll take this one

File: client/app/startup/ExampleServerReduxApp.jsx (hunk 11 above)
Comment (Gnito) [inline client/app/startup/ExampleServerReduxApp.jsx]: I'll take this one

**Comment by kpuputti** [comment]
I'm checking the tests.

**Comment by mporkola** [comment]
We decided to ditch SASS and instead use PostCSS plugins for variables, modules, mixins etc. However as it requires some effort, we also decided to postpone that work after this PR. The amount of SCSS that would potentially need altering is still quite low, and we should keep it as non-SASS as possible to facilitate the change.

**Pull request closed without merging**